{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Kunkel sequence verifier \n",
    "\n",
    "Will attempt to verify your Kunkel sequencing results and pick clones with desired mutations. \n",
    "\n",
    "Requirements so far: \n",
    "\n",
    "+ EMBOSS `merger`, for aligning two Sanger reads\n",
    "+ `blastx`, for searching protein sequences with translated nucleotide sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H09_U5271BB220-61_A7-1_T7-Ter.ab1\r\n",
      "H09_U5271BB220-61_A7-1_T7-Ter.seq\r\n",
      "H10_U5271BB220-27_A9-3_T7.ab1\r\n",
      "H10_U5271BB220-27_A9-3_T7.seq\r\n",
      "H10_U5271BB220-69_A9-3_T7-Ter.ab1\r\n",
      "H10_U5271BB220-69_A9-3_T7-Ter.seq\r\n",
      "H11_U5271BB220-35_A12-2_T7.ab1\r\n",
      "H11_U5271BB220-35_A12-2_T7.seq\r\n",
      "H11_U5271BB220-77_A12-2_T7-Ter.ab1\r\n",
      "H11_U5271BB220-77_A12-2_T7-Ter.seq\r\n"
     ]
    }
   ],
   "source": [
    "! ls 191641_dna_seq_rep_8852898_zip/ | tail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['191641_dna_seq_rep_8852898_zip/A08_U5271BB220-4_A2-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/A08_U5271BB220-46_A2-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/A09_U5271BB220-12_A4-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/A09_U5271BB220-54_A4-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/A10_U5271BB220-20_A7-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/A10_U5271BB220-62_A7-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/A11_U5271BB220-28_A10-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/A11_U5271BB220-70_A10-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/A12_U5271BB220-36_A12-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/A12_U5271BB220-78_A12-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B08_U5271BB220-5_A2-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B08_U5271BB220-47_A2-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B09_U5271BB220-13_A5-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B09_U5271BB220-55_A5-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B10_U5271BB220-21_A7-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B10_U5271BB220-63_A7-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B11_U5271BB220-29_A10-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B11_U5271BB220-71_A10-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B12_U5271BB220-37_B1-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/B12_U5271BB220-79_B1-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C08_U5271BB220-6_A2-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C08_U5271BB220-48_A2-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C09_U5271BB220-14_A5-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C09_U5271BB220-56_A5-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C10_U5271BB220-22_A8-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C10_U5271BB220-64_A8-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C11_U5271BB220-30_A10-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C11_U5271BB220-72_A10-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C12_U5271BB220-38_B1-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/C12_U5271BB220-80_B1-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D08_U5271BB220-7_A3-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D08_U5271BB220-49_A3-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D09_U5271BB220-15_A5-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D09_U5271BB220-57_A5-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D10_U5271BB220-23_A8-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D10_U5271BB220-65_A8-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D11_U5271BB220-31_A11-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D11_U5271BB220-73_A11-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D12_U5271BB220-39_B1-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/D12_U5271BB220-81_B1-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E08_U5271BB220-8_A3-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E08_U5271BB220-50_A3-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E09_U5271BB220-16_A6-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E09_U5271BB220-58_A6-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E10_U5271BB220-24_A8-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E10_U5271BB220-66_A8-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E11_U5271BB220-32_A11-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E11_U5271BB220-74_A11-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E12_U5271BB220-40_B2-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/E12_U5271BB220-82_B2-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F07_U5271BB220-1_A1-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F07_U5271BB220-43_A1-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F08_U5271BB220-9_A3-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F08_U5271BB220-51_A3-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F09_U5271BB220-17_A6-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F09_U5271BB220-59_A6-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F10_U5271BB220-25_A9-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F10_U5271BB220-67_A9-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F11_U5271BB220-33_A11-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F11_U5271BB220-75_A11-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F12_U5271BB220-41_B2-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/F12_U5271BB220-83_B2-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G07_U5271BB220-2_A1-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G07_U5271BB220-44_A1-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G08_U5271BB220-10_A4-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G08_U5271BB220-52_A4-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G09_U5271BB220-18_A6-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G09_U5271BB220-60_A6-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G10_U5271BB220-26_A9-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G10_U5271BB220-68_A9-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G11_U5271BB220-34_A12-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G11_U5271BB220-76_A12-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G12_U5271BB220-42_B2-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/G12_U5271BB220-84_B2-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H07_U5271BB220-3_A1-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H07_U5271BB220-45_A1-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H08_U5271BB220-11_A4-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H08_U5271BB220-53_A4-2_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H09_U5271BB220-19_A7-1_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H09_U5271BB220-61_A7-1_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H10_U5271BB220-27_A9-3_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H10_U5271BB220-69_A9-3_T7-Ter.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H11_U5271BB220-35_A12-2_T7.ab1']\n",
      "['191641_dna_seq_rep_8852898_zip/H11_U5271BB220-77_A12-2_T7-Ter.ab1']\n",
      "||| mutations\n",
      "['H95L'] 191641_dna_seq_rep_8852898_zip/A10.fa\n",
      "['L108N'] 191641_dna_seq_rep_8852898_zip/A11.fa\n",
      "['L290S'] 191641_dna_seq_rep_8852898_zip/A12.fa\n",
      "|||||||||||||||||||||||||||||||||||||||||||||||||| mutations\n",
      "||||||||||||||||||||||||| mutations\n",
      "||| mutations\n",
      "|||||||||||||||||||||||||||||||||||||||||||||| mutations\n",
      "|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| mutations\n",
      "['Q14H'] 191641_dna_seq_rep_8852898_zip/C08.fa\n",
      "|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| mutations\n",
      "||||||||||||||||||||||||||||||||||||| mutations\n",
      "||| mutations\n",
      "['I4X'] 191641_dna_seq_rep_8852898_zip/D08.fa\n",
      "|||||||||||||||||||||| mutations\n",
      "||| mutations\n",
      "||||||||||| mutations\n",
      "||||||||||||||||||||| mutations\n",
      "['I4X'] 191641_dna_seq_rep_8852898_zip/E08.fa\n",
      "|||||||||||| mutations\n",
      "|||||||||||||||||||||||||||||||||||| mutations\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| mutations\n",
      "|| mutations\n",
      "['Q313E'] 191641_dna_seq_rep_8852898_zip/F07.fa\n",
      "|| mutations\n",
      "|||||||||||||||||||||||||||||||||||| mutations\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/F10.fa\n",
      "||||||| mutations\n",
      "||||||||||||||||||||||||||||||||||||||||| mutations\n",
      "||||||| mutations\n",
      "|||||||| mutations\n",
      "['D322A'] 191641_dna_seq_rep_8852898_zip/G11.fa\n",
      "['H178N'] 191641_dna_seq_rep_8852898_zip/G12.fa\n",
      "|||||||||||||||||||||||||||| mutations\n",
      "||| mutations\n",
      "||||||| mutations\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/H10.fa\n",
      "||||||||| mutations\n"
     ]
    }
   ],
   "source": [
    "from itertools import product \n",
    "from subprocess import call, check_output\n",
    "from glob import glob \n",
    "\n",
    "# input\n",
    "read_dir = '191641_dna_seq_rep_8852898_zip/'\n",
    "\n",
    "# first, we need a way to quickly generate all the possible wells in a 96-well plate \n",
    "wells_human = [ '{}{:02}'.format( i, j ) for ( i, j ) in product( 'ABCDEFGH', range( 1, 13 ) ) ]\n",
    "# woo one-liners \n",
    "\n",
    "# merge \n",
    "print '{} reads total'.format( \n",
    "for well in wells_human:\n",
    "    fwd = glob( '{}/{}_*T7.ab1'.format( read_dir, well ) )\n",
    "    rev = glob( '{}/{}_*T7-Ter.ab1'.format( read_dir, well ) )\n",
    "    if fwd and rev:\n",
    "        cmd = [ 'merger', '-asequence', fwd[0], '-bsequence', rev[0], '-sreverse2' ]\n",
    "        outs = [ '-outfile', '{}/{}.merger'.format( read_dir, well ), '-outseq', '{}/{}.fa'.format( read_dir, well ) ]\n",
    "        call( cmd + outs )\n",
    "\n",
    "# align and diff \n",
    "for fa in glob( '{}/*fa'.format( read_dir ) ):\n",
    "    lines = check_output( [ 'blastx', '-subject', 'example/bglb.pep', '-query', fa, '-outfmt',  '6 sseq qseq' ] ).split()\n",
    "    z = zip( lines[0], lines[1] )\n",
    "    d = [ '{}{}{}'.format( i, n+1, j ) for n, ( i, j ) in enumerate( z ) if i != j ] \n",
    "    if len( d ) == 1:\n",
    "        print d, fa \n",
    "    elif len( d ) > 1:\n",
    "        print len( d ) * '|', 'mutations' #d, fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using phred scores to set quality threshold "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|||\n",
      "['V92D'] 191641_dna_seq_rep_8852898_zip/A09.fa\n",
      "|||\n",
      "['L108N'] 191641_dna_seq_rep_8852898_zip/A11.fa\n",
      "||\n",
      "|||\n",
      "['W325K'] 191641_dna_seq_rep_8852898_zip/B09.fa\n",
      "|||\n",
      "|||\n",
      "|||||||||||||||\n",
      "||\n",
      "||||||||||||\n",
      "['L108N'] 191641_dna_seq_rep_8852898_zip/C11.fa\n",
      "|||\n",
      "['I4X'] 191641_dna_seq_rep_8852898_zip/D08.fa\n",
      "|||\n",
      "||||||\n",
      "|||||||||||\n",
      "['I4X'] 191641_dna_seq_rep_8852898_zip/E08.fa\n",
      "['N220X'] 191641_dna_seq_rep_8852898_zip/E09.fa\n",
      "|||||||||||||||||||||||||||||||||||||||||||||||\n",
      "['W180H'] 191641_dna_seq_rep_8852898_zip/E11.fa\n",
      "||||\n",
      "['Q313E'] 191641_dna_seq_rep_8852898_zip/F07.fa\n",
      "||\n",
      "||||||||||||||||||||||||||||||\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/F10.fa\n",
      "||\n",
      "['H178N'] 191641_dna_seq_rep_8852898_zip/F12.fa\n",
      "['V311D'] 191641_dna_seq_rep_8852898_zip/G08.fa\n",
      "['N220X'] 191641_dna_seq_rep_8852898_zip/G09.fa\n",
      "['D322A'] 191641_dna_seq_rep_8852898_zip/G11.fa\n",
      "['H178N'] 191641_dna_seq_rep_8852898_zip/G12.fa\n",
      "['Q313E'] 191641_dna_seq_rep_8852898_zip/H07.fa\n",
      "['V311D'] 191641_dna_seq_rep_8852898_zip/H08.fa\n",
      "|||||||\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/H10.fa\n",
      "|||||||||\n"
     ]
    }
   ],
   "source": [
    "from itertools import product \n",
    "from subprocess import call, check_output\n",
    "from glob import glob \n",
    "from Bio import SeqIO\n",
    "\n",
    "# input\n",
    "read_dir = '191641_dna_seq_rep_8852898_zip/'\n",
    "\n",
    "# first, we need a way to quickly generate all the possible wells in a 96-well plate \n",
    "wells_human = [ '{}{:02}'.format( i, j ) for ( i, j ) in product( 'ABCDEFGH', range( 1, 13 ) ) ]\n",
    "# woo one-liners \n",
    "\n",
    "# merge \n",
    "for well in wells_human:\n",
    "    \n",
    "    fwd = glob( '{}/{}_*T7.ab1'.format( read_dir, well ) )\n",
    "    rev = glob( '{}/{}_*T7-Ter.ab1'.format( read_dir, well ) )\n",
    "\n",
    "    if fwd and rev:\n",
    "        \n",
    "        phred_qual = 10\n",
    "        \n",
    "        f = SeqIO.read( open( fwd[0], 'rb' ), \"abi\" )\n",
    "        fz = zip( f.letter_annotations['phred_quality'], f.seq )\n",
    "        fn = ''.join( [ base if qual > phred_qual else 'N' for (qual, base) in fz ] )\n",
    "        with open( '{}/{}.fwd.fasta'.format( read_dir, well ), 'w' ) as fwd_out:\n",
    "            fwd_out.write( '>{}\\n{}\\n'.format( well, fn ) )\n",
    "        \n",
    "        r = SeqIO.read( open( rev[0], 'rb' ), \"abi\" )\n",
    "        rz = zip( r.letter_annotations['phred_quality'], r.seq )\n",
    "        rn = ''.join( [ base if qual > phred_qual else 'N' for (qual, base) in rz ] )\n",
    "        with open( '{}/{}.rev.fasta'.format( read_dir, well ), 'w' ) as rev_out:\n",
    "            rev_out.write( '>{}\\n{}\\n'.format( well, rn ) )\n",
    "            \n",
    "        cmd = [ 'merger', '-asequence', '{}/{}.fwd.fasta'.format( read_dir, well ), '-bsequence', '{}/{}.rev.fasta'.format( read_dir, well ), '-sreverse2' ]\n",
    "        outs = [ '-outfile', '{}/{}.merger'.format( read_dir, well ), '-outseq', '{}/{}.fa'.format( read_dir, well ) ]\n",
    "        call( cmd + outs )\n",
    "\n",
    "# align and diff \n",
    "for fa in glob( '{}/*fa'.format( read_dir ) ):\n",
    "    lines = check_output( [ 'blastx', '-subject', 'example/bglb.pep', '-query', fa, '-outfmt',  '6 sseq qseq' ] ).split()\n",
    "    z = zip( lines[0], lines[1] )\n",
    "    d = [ '{}{}{}'.format( i, n+1, j ) for n, ( i, j ) in enumerate( z ) if i != j ] \n",
    "    if len( d ) == 1:\n",
    "        print d, fa \n",
    "    elif len( d ) > 1:\n",
    "        print len( d ) * '|' #d, fa\n",
    "                          "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
