{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Kunkel sequence verifier \n",
    "\n",
    "Will attempt to verify your Kunkel sequencing results and pick clones with desired mutations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H09_U5271BB220-61_A7-1_T7-Ter.ab1\r\n",
      "H09_U5271BB220-61_A7-1_T7-Ter.seq\r\n",
      "H10_U5271BB220-27_A9-3_T7.ab1\r\n",
      "H10_U5271BB220-27_A9-3_T7.seq\r\n",
      "H10_U5271BB220-69_A9-3_T7-Ter.ab1\r\n",
      "H10_U5271BB220-69_A9-3_T7-Ter.seq\r\n",
      "H11_U5271BB220-35_A12-2_T7.ab1\r\n",
      "H11_U5271BB220-35_A12-2_T7.seq\r\n",
      "H11_U5271BB220-77_A12-2_T7-Ter.ab1\r\n",
      "H11_U5271BB220-77_A12-2_T7-Ter.seq\r\n"
     ]
    }
   ],
   "source": [
    "! ls 191641_dna_seq_rep_8852898_zip/ | tail"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using `pear` to join paired reads \n",
    "\n",
    "Unfortunatly requires converting back and forth between file types "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw output (||| indicates 3 mutations present at amino acid level):\n",
      "['V311A'] 191641_dna_seq_rep_8852898_zip/A08.assembled.fastq\n",
      "||||||||| mutations 191641_dna_seq_rep_8852898_zip/A10.assembled.fastq\n",
      "['L108N'] 191641_dna_seq_rep_8852898_zip/A11.assembled.fastq\n",
      "['D322A'] 191641_dna_seq_rep_8852898_zip/A12.assembled.fastq\n",
      "||||||||||| mutations 191641_dna_seq_rep_8852898_zip/B08.assembled.fastq\n",
      "['W325K'] 191641_dna_seq_rep_8852898_zip/B09.assembled.fastq\n",
      "||||||| mutations 191641_dna_seq_rep_8852898_zip/B10.assembled.fastq\n",
      "||| mutations 191641_dna_seq_rep_8852898_zip/B11.assembled.fastq\n",
      "||||||||| mutations 191641_dna_seq_rep_8852898_zip/B12.assembled.fastq\n",
      "|||||| mutations 191641_dna_seq_rep_8852898_zip/C08.assembled.fastq\n",
      "|||||||||||||| mutations 191641_dna_seq_rep_8852898_zip/C09.assembled.fastq\n",
      "[] wild type 191641_dna_seq_rep_8852898_zip/C10.assembled.fastq\n",
      "|||| mutations 191641_dna_seq_rep_8852898_zip/C11.assembled.fastq\n",
      "['M323T'] 191641_dna_seq_rep_8852898_zip/C12.assembled.fastq\n",
      "[] wild type 191641_dna_seq_rep_8852898_zip/D08.assembled.fastq\n",
      "|| mutations 191641_dna_seq_rep_8852898_zip/D09.assembled.fastq\n",
      "['L272N'] 191641_dna_seq_rep_8852898_zip/D10.assembled.fastq\n",
      "||||| mutations 191641_dna_seq_rep_8852898_zip/D11.assembled.fastq\n",
      "['Q124H'] 191641_dna_seq_rep_8852898_zip/E08.assembled.fastq\n",
      "['N220K'] 191641_dna_seq_rep_8852898_zip/E09.assembled.fastq\n",
      "||||||||||||||| mutations 191641_dna_seq_rep_8852898_zip/E11.assembled.fastq\n",
      "['Q313E'] 191641_dna_seq_rep_8852898_zip/F07.assembled.fastq\n",
      "|| mutations 191641_dna_seq_rep_8852898_zip/F08.assembled.fastq\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/F10.assembled.fastq\n",
      "['W399H'] 191641_dna_seq_rep_8852898_zip/F11.assembled.fastq\n",
      "['H178N'] 191641_dna_seq_rep_8852898_zip/F12.assembled.fastq\n",
      "||| mutations 191641_dna_seq_rep_8852898_zip/G07.assembled.fastq\n",
      "||| mutations 191641_dna_seq_rep_8852898_zip/G08.assembled.fastq\n",
      "['N220Y'] 191641_dna_seq_rep_8852898_zip/G09.assembled.fastq\n",
      "[] wild type 191641_dna_seq_rep_8852898_zip/G10.assembled.fastq\n",
      "['D322A'] 191641_dna_seq_rep_8852898_zip/G11.assembled.fastq\n",
      "['H178N'] 191641_dna_seq_rep_8852898_zip/G12.assembled.fastq\n",
      "['Q313E'] 191641_dna_seq_rep_8852898_zip/H07.assembled.fastq\n",
      "['V311D'] 191641_dna_seq_rep_8852898_zip/H08.assembled.fastq\n",
      "||||||| mutations 191641_dna_seq_rep_8852898_zip/H09.assembled.fastq\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/H10.assembled.fastq\n",
      "|||||||| mutations 191641_dna_seq_rep_8852898_zip/H11.assembled.fastq\n",
      "Single mutants:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutant</th>\n",
       "      <th>well</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>V311A</td>\n",
       "      <td>A08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L108N</td>\n",
       "      <td>A11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>D322A</td>\n",
       "      <td>A12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>W325K</td>\n",
       "      <td>B09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M323T</td>\n",
       "      <td>C12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>L272N</td>\n",
       "      <td>D10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Q124H</td>\n",
       "      <td>E08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>N220K</td>\n",
       "      <td>E09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q313E</td>\n",
       "      <td>F07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>N293K</td>\n",
       "      <td>F10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>W399H</td>\n",
       "      <td>F11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>H178N</td>\n",
       "      <td>F12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>N220Y</td>\n",
       "      <td>G09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>D322A</td>\n",
       "      <td>G11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>H178N</td>\n",
       "      <td>G12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Q313E</td>\n",
       "      <td>H07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>V311D</td>\n",
       "      <td>H08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>N293K</td>\n",
       "      <td>H10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mutant well\n",
       "0   V311A  A08\n",
       "1   L108N  A11\n",
       "2   D322A  A12\n",
       "3   W325K  B09\n",
       "4   M323T  C12\n",
       "5   L272N  D10\n",
       "6   Q124H  E08\n",
       "7   N220K  E09\n",
       "8   Q313E  F07\n",
       "9   N293K  F10\n",
       "10  W399H  F11\n",
       "11  H178N  F12\n",
       "12  N220Y  G09\n",
       "13  D322A  G11\n",
       "14  H178N  G12\n",
       "15  Q313E  H07\n",
       "16  V311D  H08\n",
       "17  N293K  H10"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from itertools import product \n",
    "from subprocess import call, check_output\n",
    "from glob import glob \n",
    "import pandas \n",
    "\n",
    "# input\n",
    "read_dir = '191641_dna_seq_rep_8852898_zip/'\n",
    "\n",
    "# well list \n",
    "wells_human = [ '{}{:02}'.format( i, j ) for ( i, j ) in product( 'ABCDEFGH', range( 1, 13 ) ) ]\n",
    "\n",
    "# convert files \n",
    "from Bio import SeqIO\n",
    "\n",
    "# merge, first need to convert abi files to fastq \n",
    "count = 0\n",
    "for well in wells_human:\n",
    "    fwd = glob( '{}/{}_*T7.ab1'.format( read_dir, well ) ) \n",
    "    rev = glob( '{}/{}_*T7-Ter.ab1'.format( read_dir, well ) )\n",
    "    if fwd and rev:\n",
    "        fwdq = fwd[0].replace( 'ab1', 'fastq' )\n",
    "        revq = rev[0].replace( 'ab1', 'fastq' )\n",
    "        SeqIO.convert( fwd[0], 'abi', fwdq, 'fastq' )\n",
    "        SeqIO.convert( rev[0], 'abi', revq, 'fastq' )\n",
    "        cmd = [ 'pear', '-f', fwdq, '-r', revq, '-o', '{}/{}'.format( read_dir, well ) ] \n",
    "        call( cmd )\n",
    "        \n",
    "# align and diff, first need to convert fastq files back to fasta  \n",
    "picks = []\n",
    "\n",
    "print 'Raw output (||| indicates 3 mutations present at amino acid level):'\n",
    "for fa in glob( '{}/*.assembled.fastq'.format( read_dir ) ):\n",
    "    fq = fa.replace( 'fastq', 'fasta' )\n",
    "    SeqIO.convert( fa, 'fastq', fq, 'fasta' )\n",
    "    lines = check_output( [ 'blastx', '-subject', 'example/bglb.pep', '-query', fq, '-outfmt',  '6 sseq qseq' ] ).split()\n",
    "    if lines and lines[0] and lines[1]:\n",
    "        z = zip( lines[0], lines[1] )\n",
    "        d = [ '{}{}{}'.format( i, n+1, j ) for n, ( i, j ) in enumerate( z ) if i != j ] \n",
    "        if len( d ) == 0: \n",
    "            print '[] wild type', fa \n",
    "        if len( d ) == 1:\n",
    "            print d, fa \n",
    "            picks.append( ( d[0], fa.split( '/' )[1].split( '.' )[0] ) )\n",
    "        elif len( d ) > 1:\n",
    "            print len( d ) * '|', 'mutations', fa  #d, fa\n",
    "\n",
    "print 'Single mutants:'\n",
    "df = pandas.DataFrame( picks, columns=[ 'mutant', 'well' ] )\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using phred scores and a fixed quality threshold "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|||\n",
      "['V92D'] 191641_dna_seq_rep_8852898_zip/A09.fa\n",
      "|||\n",
      "['L108N'] 191641_dna_seq_rep_8852898_zip/A11.fa\n",
      "||\n",
      "|||\n",
      "['W325K'] 191641_dna_seq_rep_8852898_zip/B09.fa\n",
      "|||\n",
      "|||\n",
      "|||||||||||||||\n",
      "||\n",
      "||||||||||||\n",
      "['L108N'] 191641_dna_seq_rep_8852898_zip/C11.fa\n",
      "|||\n",
      "['I4X'] 191641_dna_seq_rep_8852898_zip/D08.fa\n",
      "|||\n",
      "||||||\n",
      "|||||||||||\n",
      "['I4X'] 191641_dna_seq_rep_8852898_zip/E08.fa\n",
      "['N220X'] 191641_dna_seq_rep_8852898_zip/E09.fa\n",
      "|||||||||||||||||||||||||||||||||||||||||||||||\n",
      "['W180H'] 191641_dna_seq_rep_8852898_zip/E11.fa\n",
      "||||\n",
      "['Q313E'] 191641_dna_seq_rep_8852898_zip/F07.fa\n",
      "||\n",
      "||||||||||||||||||||||||||||||\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/F10.fa\n",
      "||\n",
      "['H178N'] 191641_dna_seq_rep_8852898_zip/F12.fa\n",
      "['V311D'] 191641_dna_seq_rep_8852898_zip/G08.fa\n",
      "['N220X'] 191641_dna_seq_rep_8852898_zip/G09.fa\n",
      "['D322A'] 191641_dna_seq_rep_8852898_zip/G11.fa\n",
      "['H178N'] 191641_dna_seq_rep_8852898_zip/G12.fa\n",
      "['Q313E'] 191641_dna_seq_rep_8852898_zip/H07.fa\n",
      "['V311D'] 191641_dna_seq_rep_8852898_zip/H08.fa\n",
      "|||||||\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/H10.fa\n",
      "|||||||||\n"
     ]
    }
   ],
   "source": [
    "from itertools import product \n",
    "from subprocess import call, check_output\n",
    "from glob import glob \n",
    "from Bio import SeqIO\n",
    "\n",
    "# input\n",
    "read_dir = '191641_dna_seq_rep_8852898_zip/'\n",
    "\n",
    "# first, we need a way to quickly generate all the possible wells in a 96-well plate \n",
    "wells_human = [ '{}{:02}'.format( i, j ) for ( i, j ) in product( 'ABCDEFGH', range( 1, 13 ) ) ]\n",
    "# woo one-liners \n",
    "\n",
    "# merge \n",
    "for well in wells_human:\n",
    "    \n",
    "    fwd = glob( '{}/{}_*T7.ab1'.format( read_dir, well ) )\n",
    "    rev = glob( '{}/{}_*T7-Ter.ab1'.format( read_dir, well ) )\n",
    "\n",
    "    if fwd and rev:\n",
    "        \n",
    "        phred_qual = 10\n",
    "        \n",
    "        f = SeqIO.read( open( fwd[0], 'rb' ), \"abi\" )\n",
    "        fz = zip( f.letter_annotations['phred_quality'], f.seq )\n",
    "        fn = ''.join( [ base if qual > phred_qual else 'N' for (qual, base) in fz ] )\n",
    "        with open( '{}/{}.fwd.fasta'.format( read_dir, well ), 'w' ) as fwd_out:\n",
    "            fwd_out.write( '>{}\\n{}\\n'.format( well, fn ) )\n",
    "        \n",
    "        r = SeqIO.read( open( rev[0], 'rb' ), \"abi\" )\n",
    "        rz = zip( r.letter_annotations['phred_quality'], r.seq )\n",
    "        rn = ''.join( [ base if qual > phred_qual else 'N' for (qual, base) in rz ] )\n",
    "        with open( '{}/{}.rev.fasta'.format( read_dir, well ), 'w' ) as rev_out:\n",
    "            rev_out.write( '>{}\\n{}\\n'.format( well, rn ) )\n",
    "            \n",
    "        cmd = [ 'merger', '-asequence', '{}/{}.fwd.fasta'.format( read_dir, well ), '-bsequence', '{}/{}.rev.fasta'.format( read_dir, well ), '-sreverse2' ]\n",
    "        outs = [ '-outfile', '{}/{}.merger'.format( read_dir, well ), '-outseq', '{}/{}.fa'.format( read_dir, well ) ]\n",
    "        call( cmd + outs )\n",
    "\n",
    "# align and diff \n",
    "for fa in glob( '{}/*fa'.format( read_dir ) ):\n",
    "    lines = check_output( [ 'blastx', '-subject', 'example/bglb.pep', '-query', fa, '-outfmt',  '6 sseq qseq' ] ).split()\n",
    "    z = zip( lines[0], lines[1] )\n",
    "    d = [ '{}{}{}'.format( i, n+1, j ) for n, ( i, j ) in enumerate( z ) if i != j ] \n",
    "    if len( d ) == 1:\n",
    "        print d, fa \n",
    "    elif len( d ) > 1:\n",
    "        print len( d ) * '|' #d, fa\n",
    "                          "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Which provides better results? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
