{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Kunkel sequence verifier \n",
    "\n",
    "Will attempt to verify your Kunkel sequencing results and pick clones with desired mutations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H04_U0977BB190-95_A11-2_T7-Ter.ab1\r\n",
      "H04_U0977BB190-95_A11-2_T7-Ter.seq\r\n",
      "H05_U0977BB190-103_A14-1_T7-Ter.ab1\r\n",
      "H05_U0977BB190-103_A14-1_T7-Ter.seq\r\n",
      "H05_U0977BB190-40_A14-1_T7.ab1\r\n",
      "H05_U0977BB190-40_A14-1_T7.seq\r\n",
      "H06_U0977BB190-111_A16-3_T7-Ter.ab1\r\n",
      "H06_U0977BB190-111_A16-3_T7-Ter.seq\r\n",
      "H06_U0977BB190-48_A16-3_T7.ab1\r\n",
      "H06_U0977BB190-48_A16-3_T7.seq\r\n"
     ]
    }
   ],
   "source": [
    "! ls U0977BB190/ | tail"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using `pear` to join paired reads \n",
    "\n",
    "Unfortunatly requires converting back and forth between file types "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw output (||| indicates 3 mutations present at amino acid level):\n",
      "['N163S'] U0977BB190/A02.assembled.fastq\n",
      "|| mutations U0977BB190/A03.assembled.fastq\n",
      "['N158E'] U0977BB190/A04.assembled.fastq\n",
      "|| mutations U0977BB190/A05.assembled.fastq\n",
      "|| mutations U0977BB190/A06.assembled.fastq\n",
      "['I244V'] U0977BB190/A07.assembled.fastq\n",
      "||| mutations U0977BB190/B02.assembled.fastq\n",
      "['N163E'] U0977BB190/B04.assembled.fastq\n",
      "[] wild type U0977BB190/B05.assembled.fastq\n",
      "|||| mutations U0977BB190/B06.assembled.fastq\n",
      "['I244V'] U0977BB190/B07.assembled.fastq\n",
      "||||||||||||| mutations U0977BB190/C01.assembled.fastq\n",
      "['R240K'] U0977BB190/C02.assembled.fastq\n",
      "['E406Q'] U0977BB190/C03.assembled.fastq\n",
      "||||||| mutations U0977BB190/C04.assembled.fastq\n",
      "['N293S'] U0977BB190/C06.assembled.fastq\n",
      "|| mutations U0977BB190/C07.assembled.fastq\n",
      "|||||| mutations U0977BB190/D01.assembled.fastq\n",
      "['R240K'] U0977BB190/D02.assembled.fastq\n",
      "||||||||| mutations U0977BB190/D03.assembled.fastq\n",
      "['N220Q'] U0977BB190/D04.assembled.fastq\n",
      "||||| mutations U0977BB190/D06.assembled.fastq\n",
      "|| mutations U0977BB190/E01.assembled.fastq\n",
      "|| mutations U0977BB190/E04.assembled.fastq\n",
      "||||||| mutations U0977BB190/E05.assembled.fastq\n",
      "['N293S'] U0977BB190/E06.assembled.fastq\n",
      "|||||||||||| mutations U0977BB190/F01.assembled.fastq\n",
      "['N293C'] U0977BB190/F02.assembled.fastq\n",
      "['F1L'] U0977BB190/F03.assembled.fastq\n",
      "['I244M'] U0977BB190/F05.assembled.fastq\n",
      "['Y166F'] U0977BB190/F06.assembled.fastq\n",
      "['N163S'] U0977BB190/G01.assembled.fastq\n",
      "[] wild type U0977BB190/G02.assembled.fastq\n",
      "['N220S'] U0977BB190/G03.assembled.fastq\n",
      "|||| mutations U0977BB190/G04.assembled.fastq\n",
      "['I244M'] U0977BB190/G05.assembled.fastq\n",
      "['N163S'] U0977BB190/H01.assembled.fastq\n",
      "['W325R'] U0977BB190/H02.assembled.fastq\n",
      "['N220S'] U0977BB190/H03.assembled.fastq\n",
      "['C167S'] U0977BB190/H04.assembled.fastq\n",
      "['N220D'] U0977BB190/H05.assembled.fastq\n",
      "[] wild type U0977BB190/H06.assembled.fastq\n",
      "Single mutants:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutant</th>\n",
       "      <th>well</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N163S</td>\n",
       "      <td>A02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>N158E</td>\n",
       "      <td>A04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I244V</td>\n",
       "      <td>A07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N163E</td>\n",
       "      <td>B04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I244V</td>\n",
       "      <td>B07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>R240K</td>\n",
       "      <td>C02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>E406Q</td>\n",
       "      <td>C03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>N293S</td>\n",
       "      <td>C06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>R240K</td>\n",
       "      <td>D02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>N220Q</td>\n",
       "      <td>D04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>N293S</td>\n",
       "      <td>E06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>N293C</td>\n",
       "      <td>F02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>F1L</td>\n",
       "      <td>F03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>I244M</td>\n",
       "      <td>F05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Y166F</td>\n",
       "      <td>F06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>N163S</td>\n",
       "      <td>G01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>N220S</td>\n",
       "      <td>G03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>I244M</td>\n",
       "      <td>G05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>N163S</td>\n",
       "      <td>H01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>W325R</td>\n",
       "      <td>H02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>N220S</td>\n",
       "      <td>H03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>C167S</td>\n",
       "      <td>H04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>N220D</td>\n",
       "      <td>H05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mutant well\n",
       "0   N163S  A02\n",
       "1   N158E  A04\n",
       "2   I244V  A07\n",
       "3   N163E  B04\n",
       "4   I244V  B07\n",
       "5   R240K  C02\n",
       "6   E406Q  C03\n",
       "7   N293S  C06\n",
       "8   R240K  D02\n",
       "9   N220Q  D04\n",
       "10  N293S  E06\n",
       "11  N293C  F02\n",
       "12    F1L  F03\n",
       "13  I244M  F05\n",
       "14  Y166F  F06\n",
       "15  N163S  G01\n",
       "16  N220S  G03\n",
       "17  I244M  G05\n",
       "18  N163S  H01\n",
       "19  W325R  H02\n",
       "20  N220S  H03\n",
       "21  C167S  H04\n",
       "22  N220D  H05"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from itertools import product \n",
    "from subprocess import call, check_output\n",
    "from glob import glob \n",
    "import pandas \n",
    "\n",
    "# input\n",
    "read_dir = 'U0977BB190/'\n",
    "\n",
    "# well list \n",
    "wells_human = [ '{}{:02}'.format( i, j ) for ( i, j ) in product( 'ABCDEFGH', range( 1, 13 ) ) ]\n",
    "\n",
    "# convert files \n",
    "from Bio import SeqIO\n",
    "\n",
    "# merge, first need to convert abi files to fastq \n",
    "count = 0\n",
    "for well in wells_human:\n",
    "    fwd = glob( '{}/{}_*T7.ab1'.format( read_dir, well ) ) \n",
    "    rev = glob( '{}/{}_*T7-Ter.ab1'.format( read_dir, well ) )\n",
    "    if fwd and rev:\n",
    "        fwdq = fwd[0].replace( 'ab1', 'fastq' )\n",
    "        revq = rev[0].replace( 'ab1', 'fastq' )\n",
    "        SeqIO.convert( fwd[0], 'abi', fwdq, 'fastq' )\n",
    "        SeqIO.convert( rev[0], 'abi', revq, 'fastq' )\n",
    "        cmd = [ 'pear', '-f', fwdq, '-r', revq, '-o', '{}/{}'.format( read_dir, well ) ] \n",
    "        call( cmd )\n",
    "        \n",
    "# align and diff, first need to convert fastq files back to fasta  \n",
    "picks = []\n",
    "\n",
    "print 'Raw output (||| indicates 3 mutations present at amino acid level):'\n",
    "for fa in glob( '{}/*.assembled.fastq'.format( read_dir ) ):\n",
    "    fq = fa.replace( 'fastq', 'fasta' )\n",
    "    SeqIO.convert( fa, 'fastq', fq, 'fasta' )\n",
    "    lines = check_output( [ 'blastx', '-subject', 'example/bglb.pep', '-query', fq, '-outfmt',  '6 sseq qseq' ] ).split()\n",
    "    if lines and lines[0] and lines[1]:\n",
    "        z = zip( lines[0], lines[1] )\n",
    "        d = [ '{}{}{}'.format( i, n+1, j ) for n, ( i, j ) in enumerate( z ) if i != j ] \n",
    "        if len( d ) == 0: \n",
    "            print '[] wild type', fa \n",
    "        if len( d ) == 1:\n",
    "            print d, fa \n",
    "            picks.append( ( d[0], fa.split( '/' )[1].split( '.' )[0] ) )\n",
    "        elif len( d ) > 1:\n",
    "            print len( d ) * '|', 'mutations', fa  #d, fa\n",
    "\n",
    "print 'Single mutants:'\n",
    "df = pandas.DataFrame( picks, columns=[ 'mutant', 'well' ] )\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using phred scores and a fixed quality threshold "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|||\n",
      "['V92D'] 191641_dna_seq_rep_8852898_zip/A09.fa\n",
      "|||\n",
      "['L108N'] 191641_dna_seq_rep_8852898_zip/A11.fa\n",
      "||\n",
      "|||\n",
      "['W325K'] 191641_dna_seq_rep_8852898_zip/B09.fa\n",
      "|||\n",
      "|||\n",
      "|||||||||||||||\n",
      "||\n",
      "||||||||||||\n",
      "['L108N'] 191641_dna_seq_rep_8852898_zip/C11.fa\n",
      "|||\n",
      "['I4X'] 191641_dna_seq_rep_8852898_zip/D08.fa\n",
      "|||\n",
      "||||||\n",
      "|||||||||||\n",
      "['I4X'] 191641_dna_seq_rep_8852898_zip/E08.fa\n",
      "['N220X'] 191641_dna_seq_rep_8852898_zip/E09.fa\n",
      "|||||||||||||||||||||||||||||||||||||||||||||||\n",
      "['W180H'] 191641_dna_seq_rep_8852898_zip/E11.fa\n",
      "||||\n",
      "['Q313E'] 191641_dna_seq_rep_8852898_zip/F07.fa\n",
      "||\n",
      "||||||||||||||||||||||||||||||\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/F10.fa\n",
      "||\n",
      "['H178N'] 191641_dna_seq_rep_8852898_zip/F12.fa\n",
      "['V311D'] 191641_dna_seq_rep_8852898_zip/G08.fa\n",
      "['N220X'] 191641_dna_seq_rep_8852898_zip/G09.fa\n",
      "['D322A'] 191641_dna_seq_rep_8852898_zip/G11.fa\n",
      "['H178N'] 191641_dna_seq_rep_8852898_zip/G12.fa\n",
      "['Q313E'] 191641_dna_seq_rep_8852898_zip/H07.fa\n",
      "['V311D'] 191641_dna_seq_rep_8852898_zip/H08.fa\n",
      "|||||||\n",
      "['N293K'] 191641_dna_seq_rep_8852898_zip/H10.fa\n",
      "|||||||||\n"
     ]
    }
   ],
   "source": [
    "from itertools import product \n",
    "from subprocess import call, check_output\n",
    "from glob import glob \n",
    "from Bio import SeqIO\n",
    "\n",
    "# input\n",
    "read_dir = '191641_dna_seq_rep_8852898_zip/'\n",
    "\n",
    "# first, we need a way to quickly generate all the possible wells in a 96-well plate \n",
    "wells_human = [ '{}{:02}'.format( i, j ) for ( i, j ) in product( 'ABCDEFGH', range( 1, 13 ) ) ]\n",
    "# woo one-liners \n",
    "\n",
    "# merge \n",
    "for well in wells_human:\n",
    "    \n",
    "    fwd = glob( '{}/{}_*T7.ab1'.format( read_dir, well ) )\n",
    "    rev = glob( '{}/{}_*T7-Ter.ab1'.format( read_dir, well ) )\n",
    "\n",
    "    if fwd and rev:\n",
    "        \n",
    "        phred_qual = 10\n",
    "        \n",
    "        f = SeqIO.read( open( fwd[0], 'rb' ), \"abi\" )\n",
    "        fz = zip( f.letter_annotations['phred_quality'], f.seq )\n",
    "        fn = ''.join( [ base if qual > phred_qual else 'N' for (qual, base) in fz ] )\n",
    "        with open( '{}/{}.fwd.fasta'.format( read_dir, well ), 'w' ) as fwd_out:\n",
    "            fwd_out.write( '>{}\\n{}\\n'.format( well, fn ) )\n",
    "        \n",
    "        r = SeqIO.read( open( rev[0], 'rb' ), \"abi\" )\n",
    "        rz = zip( r.letter_annotations['phred_quality'], r.seq )\n",
    "        rn = ''.join( [ base if qual > phred_qual else 'N' for (qual, base) in rz ] )\n",
    "        with open( '{}/{}.rev.fasta'.format( read_dir, well ), 'w' ) as rev_out:\n",
    "            rev_out.write( '>{}\\n{}\\n'.format( well, rn ) )\n",
    "            \n",
    "        cmd = [ 'merger', '-asequence', '{}/{}.fwd.fasta'.format( read_dir, well ), '-bsequence', '{}/{}.rev.fasta'.format( read_dir, well ), '-sreverse2' ]\n",
    "        outs = [ '-outfile', '{}/{}.merger'.format( read_dir, well ), '-outseq', '{}/{}.fa'.format( read_dir, well ) ]\n",
    "        call( cmd + outs )\n",
    "\n",
    "# align and diff \n",
    "for fa in glob( '{}/*fa'.format( read_dir ) ):\n",
    "    lines = check_output( [ 'blastx', '-subject', 'example/bglb.pep', '-query', fa, '-outfmt',  '6 sseq qseq' ] ).split()\n",
    "    z = zip( lines[0], lines[1] )\n",
    "    d = [ '{}{}{}'.format( i, n+1, j ) for n, ( i, j ) in enumerate( z ) if i != j ] \n",
    "    if len( d ) == 1:\n",
    "        print d, fa \n",
    "    elif len( d ) > 1:\n",
    "        print len( d ) * '|' #d, fa\n",
    "                          "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Which provides better results? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
